theories/algebra/IntDiv.ec

```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
 by rewrite {1}(divz_eq (i %% k^n) (k^p)) modz_dvd_pow.
 qed.
 
-lemma nosmt modz_pow2_div n p i: 0 <= i => 0 <= p <= n =>
-  (i %% 2^n) %/ 2^p = (i %/ 2^p) %% 2^(n-p).
-proof. admitted.
+lemma nosmt divzMr i a b :  
+  0 <= a => 0 <= b => i %/ (a * b) = i %/ a %/ b.
+proof.
+move=> ge0_a ge0_b.
+case (a = 0); first by smt(div0z). 
+case (b = 0) => [| neq0_b neq0_a]; first by smt(div0z).
+pose ab := a * b; move: (edivzP i ab) (divz_eq i a) (divz_eq (i %/ a) b) => [].
+rewrite mulf_eq0 neq0_b neq0_a /= => eqi_ab rngi_ab eqi_a eqia_b.
+move: (euclideU ab (i %/ ab) (i %/ a %/ b) (i %% ab) (i %/ a %% b * a + i %% a)).
+have ->: i %/ a %/ b * ab + (i %/ a %% b * a + i %% a) = i.
++ by rewrite /ab (mulrC a) mulrA addrA -mulrDl -eqia_b -eqi_a.
+rewrite -eqi_ab rngi_ab andaE /=.
+have -> //: 0 <= i %/ a %% b * a + i %% a /\ i %/ a %% b * a + i %% a < `|ab|.
```

Comment:
Here, you can avoid the `have -> //: <some long-winded formula that's already in the goal>` with `move=> /(_ _) 2:-> //.`

`move=> /(_ _)` applies the top assumption to an anonymous proof of its first premise, requesting that proof as the first subgoal, and leaving the second subgoal with the top assumption replaced with the conclusion of that application.

Reply:
Wow, I didn't know that; that's a cool feature. And, apparently, it also works with even more underscores if your top assumption is a series of more implications; really nice!

Reply:
You don't even need this here. You can simply `move=> -> //`. EasyCrypt will do the required applications until it finds an equality to rewrite.

Reply:
Hmm, maybe I am doing it wrong, but I am having trouble achieving the same with `move=> -> //`. I think that it is maybe due to the conclusion of the top assumption being a conjunction of which I only need to rewrite with the left operand? That is, at the point where I use `move=> /(_ _ _ _) //` now, the goal is:
```
(i %/ ab * ab + i %% ab = i %/ a %/ b * ab + (i %/ a %% b * a + i %% a) =>
 0 <= i %% ab && i %% ab < `|ab| =>
 0 <= i %/ a %% b * a + i %% a && i %/ a %% b * a + i %% a < `|ab| =>
 i %/ ab = i %/ a %/ b /\ i %% ab = i %/ a %% b * a + i %% a) =>
i %/ ab = i %/ a %/ b
```
Using `move => ->` gives an error here. Even (for testing purposes) changing the `/\` to a `=>` does not allow me to use `move => -> //`. However, when the thing I want to rewrite is the actual (full) conclusion of the top assumption, it does work. I.e., if I change the above to the following, `move=> -> //` works:
```
(i %/ ab * ab + i %% ab = i %/ a %/ b * ab + (i %/ a %% b * a + i %% a) =>
 0 <= i %% ab && i %% ab < `|ab| =>
 0 <= i %/ a %% b * a + i %% a && i %/ a %% b * a + i %% a < `|ab| =>
 i %/ ab = i %/ a %/ b) =>
i %/ ab = i %/ a %/ b
```

Reply:
Your understanding is correct: `rewrite` wants a conclusion that is an equality (or an equivalence), and instead finds a conjunction. It then interprets that conjunction as being equal to `true` and tries to rewrite that, but that also fails.

@strub perhaps we could think of an intro pattern that rewrites conjunctions of equalities by (tentatively) rewriting each of the conjuncts?

Reply:
Yes, that would mimic the behavior of ssr when it encounters a tuple of equalities.